File: Assets/My Assets/Scripts/UI/MusicInstance.js
@script RequireComponent(AudioSource)
var myClips :AudioClip[];
var myNames : String[];
private var mySource : AudioSource;
var lastClipIndex : int = 0;
var fade : boolean = false;
var updatedInfo : boolean = false;
var currentIndex : int = 0;
var skiptrack : boolean = false;
function Start(){
	mySource = this.GetComponent(AudioSource);
}
function musicPlaying(){
	if(mySource.isPlaying == true){
		return true;
	}
	else{
		return false;
	}
}
function Deactivate(){
	fade = true;
}
function Activate(){
	fade = false;
}
function pickRandomClip(){
	var rand;
	if(myClips.Length <= 1){
		rand = Random.Range(0,myClips.Length+1);
	}
	else{
		rand = 0;
	}
	if(rand != lastClipIndex){
		lastClipIndex = rand;
		currentIndex = rand;
		return myClips[rand];
	}
	else{
		rand = Random.Range(0,myClips.Length+1);
		currentIndex = rand;
		return myClips[rand];	
	}
}
function updateMusicInfo(){
	var myInfo = GameObject.FindGameObjectWithTag("MusicInfo");
	if(myInfo != null){
		myInfo.GetComponent(UI.Text).text = myNames[currentIndex];
		myInfo.GetComponent("MusicInfo").BroadcastMessage("startPopUp");
	}
}
function Update(){
	if(Input.GetButtonDown("Skip")){
		if(fade == false){
			skiptrack = true;
		}
	}
	if(Application.isLoadingLevel){
		Debug.Log("Loading level! Fading music");
		fade = true;
	}
}
function LateUpdate(){
	if(musicPlaying() == false||skiptrack == true){
		mySource.clip = pickRandomClip();
		mySource.Play();
		updatedInfo = false;
		skiptrack = false;
	}
	if(fade == true){
		if(mySource.volume >= 0){
			mySource.volume-=Time.deltaTime;
		}
	}
	else{
		if(updatedInfo == false){
			updateMusicInfo();
			updatedInfo = true;
		}
		if(mySource.volume < PlayerPrefs.GetFloat("musicVolume")){
			mySource.volume+=Time.deltaTime;
		}
		if(mySource.volume > PlayerPrefs.GetFloat("musicVolume")){
			mySource.volume-=Time.deltaTime;
			if(mySource.volume - PlayerPrefs.GetFloat("musicVolume") <= 0.1){
				mySource.volume = PlayerPrefs.GetFloat("musicVolume"); //This prevents jitter
			}
		}
	}
}

File: Assets/My Assets/Scripts/UI/MusicInstance.cs
using UnityEngine;
using System.Collections;

[System.Serializable]
[UnityEngine.RequireComponent(typeof(AudioSource))]
public partial class MusicInstance : MonoBehaviour
{
    public AudioClip[] myClips;
    public string[] myNames;
    private AudioSource mySource;
    public int lastClipIndex;
    public bool fade;
    public bool updatedInfo;
    public int currentIndex;
    public bool skiptrack;
    public virtual void Start()
    {
        this.mySource = (AudioSource) this.GetComponent(typeof(AudioSource));
    }

    public virtual bool musicPlaying()
    {
        if (this.mySource.isPlaying == true)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    public virtual void Deactivate()
    {
        this.fade = true;
    }

    public virtual void Activate()
    {
        this.fade = false;
    }

    public virtual AudioClip pickRandomClip()
    {
        object rand = null;
        if (this.myClips.Length <= 1)
        {
            rand = Random.Range(0, this.myClips.Length + 1);
        }
        else
        {
            rand = 0;
        }
        if (!(rand == this.lastClipIndex))
        {
            this.lastClipIndex = (int) rand;
            this.currentIndex = (int) rand;
            return this.myClips[rand];
        }
        else
        {
            rand = Random.Range(0, this.myClips.Length + 1);
            this.currentIndex = (int) rand;
            return this.myClips[rand];
        }
    }

    public virtual void updateMusicInfo()
    {
        GameObject myInfo = GameObject.FindGameObjectWithTag("MusicInfo");
        if (myInfo != null)
        {
            myInfo.GetComponent(UI.Text).text = this.myNames[this.currentIndex];
            myInfo.GetComponent("MusicInfo").BroadcastMessage("startPopUp");
        }
    }

    public virtual void Update()
    {
        if (Input.GetButtonDown("Skip"))
        {
            if (this.fade == false)
            {
                this.skiptrack = true;
            }
        }
        if (Application.isLoadingLevel)
        {
            Debug.Log("Loading level! Fading music");
            this.fade = true;
        }
    }

    public virtual void LateUpdate()
    {
        if ((this.musicPlaying() == false) || (this.skiptrack == true))
        {
            this.mySource.clip = this.pickRandomClip();
            this.mySource.Play();
            this.updatedInfo = false;
            this.skiptrack = false;
        }
        if (this.fade == true)
        {
            if (this.mySource.volume >= 0)
            {
                this.mySource.volume = this.mySource.volume - Time.deltaTime;
            }
        }
        else
        {
            if (this.updatedInfo == false)
            {
                this.updateMusicInfo();
                this.updatedInfo = true;
            }
            if (this.mySource.volume < PlayerPrefs.GetFloat("musicVolume"))
            {
                this.mySource.volume = this.mySource.volume + Time.deltaTime;
            }
            if (this.mySource.volume > PlayerPrefs.GetFloat("musicVolume"))
            {
                this.mySource.volume = this.mySource.volume - Time.deltaTime;
                if ((this.mySource.volume - PlayerPrefs.GetFloat("musicVolume")) <= 0.1f)
                {
                    this.mySource.volume = PlayerPrefs.GetFloat("musicVolume"); //This prevents jitter
                }
            }
        }
    }

}